x-pack/plugin/autoscaling/src/internalClusterTest/java/org/elasticsearch/xpack/autoscaling/DedicatedMasterNodesDeciderIT.java

```diff
@@ -0,0 +1,327 @@
+            assertThat(deciderResults.requiredCapacity().total().memory(), equalTo(ByteSizeValue.ofGb(0)));
+        });
+
+        final ActionFuture<Void> secondRoundOfNodeStatsRequested = nodeStatsTestClient.respond(
+            (req, listener) -> respondWithFakeNodeMemory(req, minDataNodeMemory, listener),
+            3
+        );
+
+        startNodes(3, DiscoveryNodeRole.DATA_HOT_NODE_ROLE, DiscoveryNodeRole.DATA_CONTENT_NODE_ROLE, DiscoveryNodeRole.MASTER_ROLE);
+        ensureStableCluster(6);
+        secondRoundOfNodeStatsRequested.get();
+
+        assertAutoscalingDeciderResults(deciderResults -> {
+            assertThat(deciderResults.requiredCapacity().node().memory(), equalTo(ByteSizeValue.ofGb(1)));
+            assertThat(deciderResults.requiredCapacity().total().memory(), equalTo(ByteSizeValue.ofGb(3)));
+
+            assertThat(deciderResults.currentCapacity().total().memory(), equalTo(ByteSizeValue.ofGb(0)));
+        });
+    }
+
+    public void testSmallClusterDoNotNeedDedicatedMasters() throws Exception {
+        final ActionFuture<Void> nodeStatsRequestedForAllNodes = nodeStatsTestClient.respond(
+            (req, listener) -> respondWithFakeNodeMemory(req, ByteSizeValue.ofGb(1), listener),
+            7
+        );
+
+        internalCluster().setBootstrapMasterNodeIndex(0);
+        startNodes(3, DiscoveryNodeRole.DATA_HOT_NODE_ROLE, DiscoveryNodeRole.DATA_CONTENT_NODE_ROLE, DiscoveryNodeRole.MASTER_ROLE);
+        startNodes(2, DiscoveryNodeRole.DATA_WARM_NODE_ROLE);
+        startNodes(2, DiscoveryNodeRole.DATA_COLD_NODE_ROLE);
+        ensureStableCluster(7);
+
+        putAutoscalingPolicy(Settings.EMPTY);
+        nodeStatsRequestedForAllNodes.get();
+
+        assertAutoscalingDeciderResults(deciderResults -> {
+            assertThat(deciderResults.requiredCapacity().node().memory(), equalTo(ByteSizeValue.ofGb(0)));
+            assertThat(deciderResults.requiredCapacity().total().memory(), equalTo(ByteSizeValue.ofGb(0)));
+        });
+    }
+
+    public void testScaleDown() throws Exception {
+        final ActionFuture<Void> nodeStatsRequestedForAllNodes = nodeStatsTestClient.respond(
+            (req, listener) -> respondWithFakeNodeMemory(req, ByteSizeValue.ofGb(64), listener),
+            6
+        );
+
+        internalCluster().setBootstrapMasterNodeIndex(0);
+        startNodes(6, DiscoveryNodeRole.DATA_HOT_NODE_ROLE, DiscoveryNodeRole.DATA_CONTENT_NODE_ROLE, DiscoveryNodeRole.MASTER_ROLE);
+        ensureStableCluster(6);
+
+        putAutoscalingPolicy(Settings.EMPTY);
+        nodeStatsRequestedForAllNodes.get();
+
+        assertAutoscalingDeciderResults(deciderResults -> {
+            assertThat(deciderResults.requiredCapacity().node().memory(), equalTo(ByteSizeValue.ofGb(1)));
+            assertThat(deciderResults.requiredCapacity().total().memory(), equalTo(ByteSizeValue.ofGb(3)));
+        });
+
+        internalCluster().stopRandomNonMasterNode();
```

Comment:
I'm not sure if we have a way to avoid scaling down in this case when the failure is transient.